src/Agent.Listener/Configuration/ConfigurationProvider.cs

```diff
@@ -0,0 +1,357 @@
+﻿using Microsoft.TeamFoundation.DistributedTask.WebApi;
+using Microsoft.VisualStudio.Services.Agent.Util;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.TeamFoundation.Common;
+using Microsoft.VisualStudio.Services.Common;
+using Microsoft.VisualStudio.Services.WebApi;
+
+namespace Microsoft.VisualStudio.Services.Agent.Listener.Configuration
+{
+    public interface IConfigurationProvider : IExtension
+    {
+        void InitConnection(IAgentServer agentServer);
+
+        string ConfigurationProviderType { get; }
+
+        string GetServerUrl(CommandSettings command);
+
+        Task<IAgentServer> TestConnectAsync(string tfsUrl, VssCredentials creds);
+
+        Task<int> GetPoolId(CommandSettings command);
+
+        Task<TaskAgent> UpdateAgentAsync(int poolId, TaskAgent agent);
+
+        Task<TaskAgent> AddAgentAsync(int poolId, TaskAgent agent);
+
+        Task DeleteAgentAsync(int agentPoolId, int agentId);
+
+        void UpdateAgentSetting(AgentSettings settings);
+    }
+
+    public abstract class ConfigurationProvider : AgentService
+    {
+        public Type ExtensionType => typeof(IConfigurationProvider);
+        protected ITerminal _term;
+        protected IAgentServer _agentServer;
+
+        public override void Initialize(IHostContext hostContext)
+        {
+            base.Initialize(hostContext);
+            _term = hostContext.GetService<ITerminal>();
+        }
+
+        protected void InitializeServerConnection(IAgentServer agentServer)
+        {
+            _agentServer = agentServer;
+        }
+        
+        protected Task<TaskAgent> UpdateAgent(int poolId, TaskAgent agent)
+        {
+           return _agentServer.UpdateAgentAsync(poolId, agent);
+        }
+
+        protected Task<TaskAgent> AddAgent(int poolId, TaskAgent agent)
+        {
+            return _agentServer.AddAgentAsync(poolId, agent);
+        }
+
+        protected Task DeleteAgent(int poolId, int agentId)
+        {
+            return _agentServer.DeleteAgentAsync(poolId, agentId);
+        }
+
+        protected async Task TestConnectionAsync(string url, VssCredentials creds)
+        {
+            _term.WriteLine(StringUtil.Loc("ConnectingToServer"));
+            VssConnection connection = ApiUtil.CreateConnection(new Uri(url), creds);
+
+            _agentServer = HostContext.CreateService<IAgentServer>();
```

Comment:
1) In case of VSTS only one IAgentServer is required and we create only one
2) In case of on-prem we required two - one at account level and other at collection level. The collection level is private one for machine group provider